<!doctype html>
<html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"><meta><title>scRNA分析流程（一） - Hat&#039;s blog</title><link rel="manifest" href="/manifest.json"><meta name="application-name" content="Hat&#039;s blog"><meta name="msapplication-TileImage" content="/imgs/dinosaur.jpg"><meta name="apple-mobile-web-app-capable" content="yes"><meta name="apple-mobile-web-app-title" content="Hat&#039;s blog"><meta name="apple-mobile-web-app-status-bar-style" content="default"><meta name="description" content="scRNA分析在最近几年发展迅速，对于其分析流程也比较成熟了。现在主流的scRNA分析主要有两种。一种是10×分析流程，一种的Smart-seq分析流程。在此我整理了一套scRNA的分析流程供大家讨论。"><meta property="og:type" content="blog"><meta property="og:title" content="scRNA分析流程（一）"><meta property="og:url" content="http://example.com/2023/11/11/scRNA_1/"><meta property="og:site_name" content="Hat&#039;s blog"><meta property="og:description" content="scRNA分析在最近几年发展迅速，对于其分析流程也比较成熟了。现在主流的scRNA分析主要有两种。一种是10×分析流程，一种的Smart-seq分析流程。在此我整理了一套scRNA的分析流程供大家讨论。"><meta property="og:locale" content="en_US"><meta property="og:image" content="http://example.com/img/og_image.png"><meta property="article:published_time" content="2023-11-11T13:15:05.386Z"><meta property="article:modified_time" content="2023-11-11T13:15:40.337Z"><meta property="article:author" content="Se"><meta property="article:tag" content="Pipeline"><meta property="article:tag" content="scRNA"><meta property="twitter:card" content="summary"><meta property="twitter:image:src" content="http://example.com/img/og_image.png"><script type="application/ld+json">{"@context":"https://schema.org","@type":"BlogPosting","mainEntityOfPage":{"@type":"WebPage","@id":"http://example.com/2023/11/11/scRNA_1/"},"headline":"scRNA分析流程（一）","image":["http://example.com/img/og_image.png"],"datePublished":"2023-11-11T13:15:05.386Z","dateModified":"2023-11-11T13:15:40.337Z","author":{"@type":"Person","name":"Se"},"publisher":{"@type":"Organization","name":"Hat's blog","logo":{"@type":"ImageObject","url":"http://example.com/imgs/dinosaur.jpg"}},"description":"scRNA分析在最近几年发展迅速，对于其分析流程也比较成熟了。现在主流的scRNA分析主要有两种。一种是10×分析流程，一种的Smart-seq分析流程。在此我整理了一套scRNA的分析流程供大家讨论。"}</script><link rel="canonical" href="http://example.com/2023/11/11/scRNA_1/"><link rel="icon" href="/imgs/dinosaur.jpg"><link rel="stylesheet" href="https://use.fontawesome.com/releases/v6.0.0/css/all.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/highlight.js@9.12.0/styles/atom-one-light.css"><link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Ubuntu:wght@400;600&amp;family=Source+Code+Pro"><link rel="stylesheet" href="/css/default.css"><style>body>.footer,body>.navbar,body>.section{opacity:0}</style><!--!--><!--!--><!--!--><!--!--><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/cookieconsent@3.1.1/build/cookieconsent.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/lightgallery@1.10.0/dist/css/lightgallery.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/justifiedGallery@3.8.1/dist/css/justifiedGallery.min.css"><!--!--><!--!--><style>.pace{-webkit-pointer-events:none;pointer-events:none;-webkit-user-select:none;-moz-user-select:none;user-select:none}.pace-inactive{display:none}.pace .pace-progress{background:#3273dc;position:fixed;z-index:2000;top:0;right:100%;width:100%;height:2px}</style><script src="https://cdn.jsdelivr.net/npm/pace-js@1.2.4/pace.min.js"></script><!--!--><!--!--><!-- hexo injector head_end start --><script>
  (function () {
      function switchTab() {
          if (!location.hash) {
            return;
          }

          const id = '#' + CSS.escape(location.hash.substring(1));
          const $tabMenu = document.querySelector(`.tabs a[href="${id}"]`);
          if (!$tabMenu) {
            return;
          }

          const $tabMenuContainer = $tabMenu.parentElement.parentElement;
          Array.from($tabMenuContainer.children).forEach($menu => $menu.classList.remove('is-active'));
          Array.from($tabMenuContainer.querySelectorAll('a'))
              .map($menu => document.getElementById($menu.getAttribute("href").substring(1)))
              .forEach($content => $content.classList.add('is-hidden'));

          if ($tabMenu) {
              $tabMenu.parentElement.classList.add('is-active');
          }
          const $activeTab = document.querySelector(id);
          if ($activeTab) {
              $activeTab.classList.remove('is-hidden');
          }
      }
      switchTab();
      window.addEventListener('hashchange', switchTab, false);
  })();
  </script><!-- hexo injector head_end end --><meta name="generator" content="Hexo 6.3.0"></head><body class="is-2-column"><nav class="navbar navbar-main"><div class="container navbar-container"><div class="navbar-brand justify-content-center"><a class="navbar-item navbar-logo" href="/"><img src="/imgs/dinosaur.jpg" alt="Hat&#039;s blog" height="28"></a></div><div class="navbar-menu"><div class="navbar-start"><a class="navbar-item" href="/">Home</a><a class="navbar-item" href="/archives">Archives</a><a class="navbar-item" href="/categories">Categories</a><a class="navbar-item" href="/about">About</a></div><div class="navbar-end"><a class="navbar-item" target="_blank" rel="noopener" title="Download on GitHub" href="https://github.com/iAMSe/"><i class="fab fa-github"></i></a><a class="navbar-item is-hidden-tablet catalogue" title="Catalogue" href="javascript:;"><i class="fas fa-list-ul"></i></a><a class="navbar-item search" title="Search" href="javascript:;"><i class="fas fa-search"></i></a></div></div></div></nav><section class="section"><div class="container"><div class="columns"><div class="column order-2 column-main is-8-tablet is-8-desktop is-8-widescreen"><div class="card"><article class="card-content article" role="article"><div class="article-meta is-size-7 is-uppercase level is-mobile"><div class="level-left"><span class="level-item">Posted&nbsp;<time dateTime="2023-11-11T13:15:05.386Z" title="2023/11/11 21:15:05">2023-11-11</time></span><span class="level-item">Updated&nbsp;<time dateTime="2023-11-11T13:15:40.337Z" title="2023/11/11 21:15:40">2023-11-11</time></span><span class="level-item"> Se </span><span class="level-item"><a class="link-muted" href="/categories/pipeline/">pipeline</a></span></div></div><h1 class="title is-3 is-size-4-mobile">scRNA分析流程（一）</h1><div class="content"><p>scRNA分析在最近几年发展迅速，对于其分析流程也比较成熟了。现在主流的scRNA分析主要有两种。一种是<a target="_blank" rel="noopener" href="https://www.10xgenomics.com/">10×</a>分析流程，一种的<a target="_blank" rel="noopener" href="https://www.illumina.com/science/sequencing-method-explorer/kits-and-arrays/smart-seq2.html">Smart-seq</a>分析流程。在此我整理了一套scRNA的分析流程供大家讨论。</p>
<span id="more"></span>



<h1 id="数据处理"><a href="#数据处理" class="headerlink" title="数据处理"></a>数据处理</h1><h2 id="Smart-seq数据"><a href="#Smart-seq数据" class="headerlink" title="Smart-seq数据"></a>Smart-seq数据</h2><p><strong>Smart-seq</strong>数据其实是通过实验方法将细胞分成单个细胞，然后再进行测序。所以对<strong>Smart-seq</strong>数据其实可以参考之前<a target="_blank" rel="noopener" href="https://iamse.github.io/2023/10/26/RNAseq/#more">RNA-seq</a>数据的处理方法。相比较<strong>Bulk RNA-seq</strong>数据，<strong>Smart-seq</strong>的质量相对来说会低一些。</p>
<h2 id="10×数据处理"><a href="#10×数据处理" class="headerlink" title="10×数据处理"></a>10×数据处理</h2><p>更高通量的单细胞数据处理方法是采用<a target="_blank" rel="noopener" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">cellrange</a>的来处理的。其中需要注意的是，<code>fastq</code>文件需要将名字修改成对应格式<code>[Sample Name]_S1_L00[Lane Number]_[Read Type]_001.fastq.gz</code>。其中<code>I1, I2</code>是样本的index号，<code>R1, R2</code>是Read1、Read2。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">/path/to/sample1_S1_L001_R1_001.fastq.gz</span><br><span class="line">/path/to/sample2_S1_L001_R2_001.fastq.gz</span><br></pre></td></tr></table></figure>



<p>此外你还需要准备好Reference。Cellranger提供了小鼠和人类的参考基因组下载。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">fastq_path=/path/to/fastq</span><br><span class="line">cellranger=/path/to/cellranger-5.0.1/cellranger</span><br><span class="line">transcriptome=/path/to/cellranger-5.0.1/cellranger/refdata-gex-GRCh38-2020-A/</span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_"></span></span><br><span class="line"><span class="meta prompt_">$</span><span class="language-bash">cellranger count --<span class="built_in">id</span>=<span class="variable">$id</span> --transcriptome=<span class="variable">$transcriptome</span> --fastqs=<span class="variable">$fastq_path</span> --sample=<span class="variable">$id</span></span></span><br></pre></td></tr></table></figure>



<p><code>--transcriptome</code>参考基因组路径</p>
<p><code>--fastqs</code>fastq文件路径</p>
<h1 id="Seurat-处理"><a href="#Seurat-处理" class="headerlink" title="Seurat 处理"></a>Seurat 处理</h1><h2 id="加载数据"><a href="#加载数据" class="headerlink" title="加载数据"></a>加载数据</h2><h3 id="Smart-seq"><a href="#Smart-seq" class="headerlink" title="Smart seq"></a>Smart seq</h3><p><strong>Smart-seq</strong>数据需要整理一个矩阵。每一行是一个基因的reads数，每一列是一个样本。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">Gene	sample1	sample2	...	sampleN</span><br><span class="line">gene1	1	14	... 204</span><br><span class="line">gene2	4	23	...	322</span><br><span class="line">...	</span><br><span class="line">geneN	34	123	...	43</span><br></pre></td></tr></table></figure>

<p>同时需要一个<strong>meta</strong>文件。每一行都是一个样本信息</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cellid	Age	CellType</span><br><span class="line">sample1	50	delta</span><br><span class="line">sample2	60	acinar</span><br><span class="line">sample3	68	ductal</span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## load data and build Seurat object</span></span><br><span class="line">cr <span class="operator">=</span> read.table<span class="punctuation">(</span> paste0<span class="punctuation">(</span>sid<span class="punctuation">,</span> <span class="string">&quot;_scRNA_matrix.gz&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> head<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span> row.names<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span>check.names<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line">cr <span class="operator">=</span> cr<span class="punctuation">[</span> <span class="punctuation">,</span> row.names<span class="punctuation">(</span>info<span class="punctuation">)</span> <span class="punctuation">]</span>;</span><br><span class="line">sce<span class="operator">=</span> CreateSeuratObject<span class="punctuation">(</span> counts<span class="operator">=</span>cr<span class="punctuation">,</span> project<span class="operator">=</span>sid<span class="punctuation">,</span> min.cells<span class="operator">=</span><span class="number">3</span><span class="punctuation">,</span> min.features<span class="operator">=</span><span class="number">100</span> <span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line"><span class="comment">## load info, select cells</span></span><br><span class="line">info <span class="operator">=</span> read.table<span class="punctuation">(</span><span class="string">&quot;cell.anno&quot;</span><span class="punctuation">,</span> row.names<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span> head<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">)</span></span><br><span class="line">anno <span class="operator">=</span> data.frame<span class="punctuation">(</span> info<span class="punctuation">[</span><span class="punctuation">,</span> <span class="string">&quot;CellType&quot;</span><span class="punctuation">]</span> <span class="punctuation">)</span></span><br><span class="line">row.names<span class="punctuation">(</span>anno<span class="punctuation">)</span> <span class="operator">=</span> row.names<span class="punctuation">(</span>info<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">age <span class="operator">=</span> data.frame<span class="punctuation">(</span> info<span class="punctuation">[</span><span class="punctuation">,</span> <span class="string">&quot;Age&quot;</span><span class="punctuation">]</span> <span class="punctuation">)</span></span><br><span class="line">row.names<span class="punctuation">(</span>age<span class="punctuation">)</span> <span class="operator">=</span> row.names<span class="punctuation">(</span>info<span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>过滤然后标准化后可以通过meta文件添加样本信息</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## load Cell annotations</span></span><br><span class="line"><span class="comment">## 3 features: CellType, Age</span></span><br><span class="line">sce <span class="operator">=</span> AddMetaData<span class="punctuation">(</span>sce<span class="punctuation">,</span> as.vector<span class="punctuation">(</span>anno<span class="punctuation">)</span><span class="punctuation">,</span> col.name<span class="operator">=</span><span class="string">&quot;CellType&quot;</span><span class="punctuation">)</span></span><br><span class="line">sce <span class="operator">=</span> AddMetaData<span class="punctuation">(</span>sce<span class="punctuation">,</span> as.vector<span class="punctuation">(</span>age<span class="punctuation">)</span><span class="punctuation">,</span> col.name<span class="operator">=</span><span class="string">&quot;Age&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="10×"><a href="#10×" class="headerlink" title="10×"></a>10×</h3><p>其实<strong>10×</strong>的数据很简单只需要<code>.h5</code>文件就好了。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">suppressPackageStartupMessages<span class="punctuation">(</span><span class="punctuation">&#123;</span></span><br><span class="line">	library<span class="punctuation">(</span>Seurat<span class="punctuation">)</span>;library<span class="punctuation">(</span>plyr<span class="punctuation">)</span>;library<span class="punctuation">(</span>dplyr<span class="punctuation">)</span>;library<span class="punctuation">(</span>cowplot<span class="punctuation">)</span>;</span><br><span class="line">	library<span class="punctuation">(</span>ggplot2<span class="punctuation">)</span>;library<span class="punctuation">(</span>gridExtra<span class="punctuation">)</span>;library<span class="punctuation">(</span>ggthemes<span class="punctuation">)</span>;library<span class="punctuation">(</span>patchwork<span class="punctuation">)</span>;</span><br><span class="line">	library<span class="punctuation">(</span>forcats<span class="punctuation">)</span>;library<span class="punctuation">(</span>DoubletFinder<span class="punctuation">)</span>;</span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span>;</span><br><span class="line">options<span class="punctuation">(</span> stringsAsFactors<span class="operator">=</span><span class="built_in">F</span> <span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line">sid <span class="operator">=</span> <span class="string">&#x27;MDS4&#x27;</span>;</span><br><span class="line"></span><br><span class="line"><span class="comment">## load data and build Seurat object</span></span><br><span class="line">cr <span class="operator">=</span> Read10X_h5<span class="punctuation">(</span> paste0<span class="punctuation">(</span>sid<span class="punctuation">,</span> <span class="string">&#x27;.h5&#x27;</span><span class="punctuation">)</span> <span class="punctuation">)</span>;</span><br><span class="line">sc.raw <span class="operator">=</span> CreateSeuratObject<span class="punctuation">(</span> counts<span class="operator">=</span>cr<span class="punctuation">,</span> project<span class="operator">=</span>sid<span class="punctuation">,</span> min.cells<span class="operator">=</span><span class="number">3</span><span class="punctuation">,</span> min.features<span class="operator">=</span><span class="number">200</span> <span class="punctuation">)</span>;</span><br></pre></td></tr></table></figure>

<h2 id="质控"><a href="#质控" class="headerlink" title="质控"></a>质控</h2><p>对于线粒体基因数数量过高的细胞（可能是细胞活性较低），基因数量过低的细胞或者基因数量过高的多胞细胞也要排除。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Quality control</span></span><br><span class="line">sc.raw<span class="punctuation">[[</span><span class="string">&quot;percent.mito&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">=</span> PercentageFeatureSet<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> pattern<span class="operator">=</span><span class="string">&quot;^MT-&quot;</span><span class="punctuation">)</span>;</span><br><span class="line">sc.raw<span class="punctuation">[[</span><span class="string">&quot;percent.ribo&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">=</span> PercentageFeatureSet<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> pattern<span class="operator">=</span><span class="string">&quot;^RP[SL]&quot;</span><span class="punctuation">)</span>;</span><br><span class="line">p1<span class="operator">=</span>VlnPlot<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> features<span class="operator">=</span><span class="string">&#x27;nCount_RNA&#x27;</span><span class="punctuation">,</span> y.max<span class="operator">=</span><span class="number">70000</span><span class="punctuation">,</span></span><br><span class="line">		   pt.size<span class="operator">=</span><span class="number">0.05</span><span class="punctuation">)</span><span class="operator">+</span>NoLegend<span class="punctuation">(</span><span class="punctuation">)</span>;</span><br><span class="line">p2<span class="operator">=</span>VlnPlot<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> features<span class="operator">=</span><span class="string">&#x27;nFeature_RNA&#x27;</span><span class="punctuation">,</span> y.max<span class="operator">=</span><span class="number">10000</span><span class="punctuation">,</span></span><br><span class="line">		   pt.size<span class="operator">=</span><span class="number">0.05</span><span class="punctuation">)</span><span class="operator">+</span>NoLegend<span class="punctuation">(</span><span class="punctuation">)</span>;</span><br><span class="line">p3<span class="operator">=</span>VlnPlot<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> features<span class="operator">=</span><span class="string">&#x27;percent.mito&#x27;</span><span class="punctuation">,</span> y.max<span class="operator">=</span><span class="number">100</span><span class="punctuation">,</span></span><br><span class="line">		   pt.size<span class="operator">=</span><span class="number">0.05</span><span class="punctuation">)</span><span class="operator">+</span>NoLegend<span class="punctuation">(</span><span class="punctuation">)</span>;</span><br><span class="line">p4<span class="operator">=</span>VlnPlot<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> features<span class="operator">=</span><span class="string">&#x27;percent.ribo&#x27;</span><span class="punctuation">,</span> y.max<span class="operator">=</span><span class="number">100</span><span class="punctuation">,</span></span><br><span class="line">		   pt.size<span class="operator">=</span><span class="number">0.05</span><span class="punctuation">)</span><span class="operator">+</span>NoLegend<span class="punctuation">(</span><span class="punctuation">)</span>;</span><br><span class="line">grid.arrange<span class="punctuation">(</span>p1<span class="punctuation">,</span> p2<span class="punctuation">,</span> p3<span class="punctuation">,</span> p4<span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span></span><br><span class="line">			 top<span class="operator">=</span>paste0<span class="punctuation">(</span><span class="string">&#x27;Before QC, cells=&#x27;</span><span class="punctuation">,</span> <span class="built_in">dim</span><span class="punctuation">(</span>sc.raw<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line">p5<span class="operator">=</span>FeatureScatter<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> feature1<span class="operator">=</span><span class="string">&quot;nCount_RNA&quot;</span><span class="punctuation">,</span> feature2<span class="operator">=</span><span class="string">&quot;percent.mito&quot;</span><span class="punctuation">)</span>;</span><br><span class="line">p6<span class="operator">=</span>FeatureScatter<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> feature1<span class="operator">=</span><span class="string">&quot;nCount_RNA&quot;</span><span class="punctuation">,</span> feature2<span class="operator">=</span><span class="string">&quot;nFeature_RNA&quot;</span><span class="punctuation">)</span>;</span><br><span class="line">p7<span class="operator">=</span>FeatureScatter<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> feature1<span class="operator">=</span><span class="string">&quot;nCount_RNA&quot;</span><span class="punctuation">,</span> feature2<span class="operator">=</span><span class="string">&quot;percent.ribo&quot;</span><span class="punctuation">)</span>;</span><br><span class="line">p8<span class="operator">=</span>FeatureScatter<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> feature1<span class="operator">=</span><span class="string">&quot;percent.mito&quot;</span><span class="punctuation">,</span> feature2<span class="operator">=</span><span class="string">&quot;percent.ribo&quot;</span><span class="punctuation">)</span>;</span><br><span class="line">grid.arrange<span class="punctuation">(</span>p5<span class="punctuation">,</span> p6<span class="punctuation">,</span> p7<span class="punctuation">,</span> p8<span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">2</span><span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line"><span class="comment">## filter, mostly filter out the cells with abnormally high chrM (i.e., dead cells)</span></span><br><span class="line"><span class="comment">## outliners should also be filtered</span></span><br><span class="line"><span class="comment">#sce = subset(sc.raw, subset=nCount_RNA&lt;4500 &amp; nFeature_RNA&lt;1500 &amp; percent.mito&lt;25);</span></span><br><span class="line">sce <span class="operator">=</span> subset<span class="punctuation">(</span>sc.raw<span class="punctuation">,</span> percent.mito<span class="operator">&lt;</span><span class="number">25</span><span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line"><span class="comment">## regress out unwanted genes</span></span><br><span class="line">all.gene <span class="operator">=</span> rownames<span class="punctuation">(</span> sce <span class="punctuation">)</span>;</span><br><span class="line">cc.genes <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span>cc.genes.updated.2019<span class="operator">$</span>s.genes<span class="punctuation">,</span> cc.genes.updated.2019<span class="operator">$</span>g2m.genes<span class="punctuation">)</span>;</span><br><span class="line">RP<span class="operator">=</span>grep<span class="punctuation">(</span>rownames<span class="punctuation">(</span>sce<span class="punctuation">)</span><span class="punctuation">,</span> pattern<span class="operator">=</span><span class="string">&quot;^RP[SL]&quot;</span><span class="punctuation">,</span></span><br><span class="line">		value<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span> invert<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span> ignore.case<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">)</span>;</span><br><span class="line">MT<span class="operator">=</span>grep<span class="punctuation">(</span>rownames<span class="punctuation">(</span>sce<span class="punctuation">)</span><span class="punctuation">,</span> pattern<span class="operator">=</span><span class="string">&quot;^MT-&quot;</span><span class="punctuation">,</span></span><br><span class="line">		value<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span> invert<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span> ignore.case<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">)</span>;</span><br><span class="line">keep.genes <span class="operator">=</span> all.gene<span class="punctuation">[</span><span class="operator">!</span><span class="punctuation">(</span>all.gene <span class="operator">%in%</span> <span class="built_in">c</span><span class="punctuation">(</span>cc.genes<span class="punctuation">,</span> RP<span class="punctuation">,</span> MT<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">]</span>;</span><br><span class="line">sce <span class="operator">=</span> subset<span class="punctuation">(</span>sce<span class="punctuation">,</span> features<span class="operator">=</span>keep.genes<span class="punctuation">)</span>;</span><br></pre></td></tr></table></figure>

<h2 id="标准化-PCA"><a href="#标准化-PCA" class="headerlink" title="标准化 PCA"></a>标准化 PCA</h2><p>一般来说我们都使用的**<code>LogNormalize</code>**方法来进行标准化</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## normalize &amp; scale data, run PCA</span></span><br><span class="line">sce <span class="operator">=</span> NormalizeData<span class="punctuation">(</span>sce<span class="punctuation">,</span> normalization.method<span class="operator">=</span><span class="string">&#x27;LogNormalize&#x27;</span><span class="punctuation">,</span></span><br><span class="line">					scale.factor<span class="operator">=</span><span class="number">10000</span><span class="punctuation">,</span> verbose<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> FindVariableFeatures<span class="punctuation">(</span>sce<span class="punctuation">,</span> selection.method<span class="operator">=</span><span class="string">&#x27;vst&#x27;</span><span class="punctuation">,</span></span><br><span class="line">					nfeatures<span class="operator">=</span><span class="number">2000</span><span class="punctuation">,</span> verbose<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> ScaleData<span class="punctuation">(</span>sce<span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> RunPCA<span class="punctuation">(</span>sce<span class="punctuation">,</span> features<span class="operator">=</span>VariableFeatures<span class="punctuation">(</span>sce<span class="punctuation">)</span><span class="punctuation">,</span> verbose<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line">sce <span class="operator">=</span> FindNeighbors<span class="punctuation">(</span>sce<span class="punctuation">,</span> dims<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">30</span><span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> FindClusters<span class="punctuation">(</span>sce<span class="punctuation">,</span> resolution<span class="operator">=</span><span class="number">0.5</span><span class="punctuation">,</span> verbose<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> RunTSNE<span class="punctuation">(</span>sce<span class="punctuation">,</span> dims<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">30</span><span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> RunUMAP<span class="punctuation">(</span>sce<span class="punctuation">,</span> dims<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">30</span><span class="punctuation">,</span> verbose<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line"><span class="comment">#DimPlot(sce,reduction = &quot;tsne&quot;, label=T,split.by =&#x27;orig.ident&#x27;);</span></span><br><span class="line"><span class="comment">#LabelClusters(DimPlot(sce, reduction = &quot;tsne&quot;),id = &#x27;ident&#x27;)</span></span><br><span class="line">UMAPPlot<span class="punctuation">(</span>sce<span class="punctuation">,</span> pt.size<span class="operator">=</span><span class="number">1</span><span class="punctuation">,</span> label<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span> group.by<span class="operator">=</span><span class="string">&#x27;RNA_snn_res.0.5&#x27;</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">ggtitle<span class="punctuation">(</span><span class="string">&#x27;res=0.5, before doublet removal&#x27;</span><span class="punctuation">)</span>;</span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>此外还提供了**<code>SCTransform()</code>**方法来进行标准化。这一个函数可以代替三个函数（NormalizeData, ScaleData, FindVariableFeatures）的运行。 且其对测序深度的校正效果要好于log标准化。（10万以内的细胞都建议使用SCT标准化）。SCTransform对测序深度的校正效果很好，也可用于矫正线粒体等因素的影响，但不能用于批次矫正。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## find the optimal parameter, runs slowly</span></span><br><span class="line"><span class="comment">## if log-normalization is used, set sct=F (default)</span></span><br><span class="line"><span class="comment">## otherwise set it to T (is SCT-normalization is used)</span></span><br><span class="line">sweep.res.list <span class="operator">=</span> paramSweep_v3<span class="punctuation">(</span>sce<span class="punctuation">,</span> PCs<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">30</span><span class="punctuation">,</span> sct<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line">sweep.stats <span class="operator">=</span> summarizeSweep<span class="punctuation">(</span>sweep.res.list<span class="punctuation">,</span> GT<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line">bcmvn <span class="operator">=</span> find.pK<span class="punctuation">(</span>sweep.stats<span class="punctuation">)</span>;</span><br><span class="line">pK_good <span class="operator">=</span> bcmvn<span class="operator">$</span>pK<span class="punctuation">[</span>which.max<span class="punctuation">(</span>bcmvn<span class="operator">$</span>BCmetric<span class="punctuation">)</span><span class="punctuation">]</span> <span class="operator">%&gt;%</span> <span class="built_in">as.character</span><span class="punctuation">(</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> <span class="built_in">as.numeric</span><span class="punctuation">(</span><span class="punctuation">)</span>;</span><br></pre></td></tr></table></figure>

<p>对双胞进一步过滤</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## in 10x, doublet rate is related to the No. of recovered cells,</span></span><br><span class="line"><span class="comment">## increases ~0.8% when recovering additional 1000 cells</span></span><br><span class="line"><span class="comment">## Note that here I use the sc.raw as the recovered cells</span></span><br><span class="line">DoubletRate <span class="operator">=</span> <span class="built_in">dim</span><span class="punctuation">(</span>sc.raw<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span> <span class="operator">*</span> <span class="number">8e-6</span>;</span><br><span class="line">homotypic.prop <span class="operator">=</span> modelHomotypic<span class="punctuation">(</span>sce<span class="operator">$</span>seurat_clusters<span class="punctuation">)</span>;</span><br><span class="line">nExp_poi <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>DoubletRate<span class="operator">*</span>ncol<span class="punctuation">(</span>sce<span class="punctuation">)</span><span class="punctuation">)</span>;</span><br><span class="line">nExp_poi.adj <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>nExp_poi<span class="operator">*</span><span class="punctuation">(</span><span class="number">1</span><span class="operator">-</span>homotypic.prop<span class="punctuation">)</span><span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> doubletFinder_v3<span class="punctuation">(</span>sce<span class="punctuation">,</span> PCs<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">30</span><span class="punctuation">,</span> pN<span class="operator">=</span><span class="number">0.25</span><span class="punctuation">,</span> pK<span class="operator">=</span>pK_good<span class="punctuation">,</span></span><br><span class="line">				nExp<span class="operator">=</span>nExp_poi.adj<span class="punctuation">,</span> reuse.pANN<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span> sct<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line"><span class="comment">## Plot results</span></span><br><span class="line"><span class="comment">## note that DoubletFinder will add 1 column named &quot;DF.classifications_XXXX&quot;</span></span><br><span class="line"><span class="comment">## which is parameter-specific and need to be renamed</span></span><br><span class="line">index  <span class="operator">=</span> grep<span class="punctuation">(</span><span class="string">&quot;DF.classifications&quot;</span><span class="punctuation">,</span> colnames<span class="punctuation">(</span>sce<span class="operator">@</span>meta.data<span class="punctuation">)</span><span class="punctuation">)</span>;</span><br><span class="line">colnames<span class="punctuation">(</span>sce<span class="operator">@</span>meta.data<span class="punctuation">)</span><span class="punctuation">[</span>index<span class="punctuation">]</span> <span class="operator">=</span> <span class="string">&quot;DFpred&quot;</span>;</span><br><span class="line">DimPlot<span class="punctuation">(</span>sce<span class="punctuation">,</span> reduction<span class="operator">=</span><span class="string">&quot;umap&quot;</span><span class="punctuation">,</span> group.by<span class="operator">=</span><span class="string">&quot;DFpred&quot;</span><span class="punctuation">,</span></span><br><span class="line">		order<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;Singlet&quot;</span><span class="punctuation">,</span> <span class="string">&quot;Doublet&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> cols<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;red&quot;</span><span class="punctuation">,</span> <span class="string">&quot;grey&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">ggtitle<span class="punctuation">(</span> paste0<span class="punctuation">(</span><span class="string">&quot;DoubletFinder result, pK=&quot;</span><span class="punctuation">,</span> pK_good<span class="punctuation">)</span> <span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line">write.table<span class="punctuation">(</span>sce<span class="operator">$</span>DFpred<span class="punctuation">,</span>	file<span class="operator">=</span>paste0<span class="punctuation">(</span>sid<span class="punctuation">,</span> <span class="string">&quot;.DoubletFinder.result&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">			<span class="built_in">quote</span><span class="operator">=</span><span class="built_in">F</span><span class="punctuation">,</span> sep<span class="operator">=</span><span class="string">&#x27;\t&#x27;</span><span class="punctuation">,</span> row.names<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span> col.names<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line"><span class="comment">## show No. of cells per cluster</span></span><br><span class="line">count.before.DF <span class="operator">=</span> as.data.frame<span class="punctuation">(</span>table<span class="punctuation">(</span>sce<span class="operator">@</span>meta.data<span class="operator">$</span>RNA_snn_res.0.5<span class="punctuation">)</span><span class="punctuation">)</span>;</span><br><span class="line">total.before.DF <span class="operator">=</span> <span class="built_in">dim</span><span class="punctuation">(</span>sce<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span>;</span><br><span class="line"></span><br><span class="line"><span class="comment">## filter out doublets</span></span><br><span class="line">sce <span class="operator">=</span> subset<span class="punctuation">(</span> sce<span class="punctuation">,</span> DFpred<span class="operator">==</span><span class="string">&quot;Singlet&quot;</span> <span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> FindNeighbors<span class="punctuation">(</span>sce<span class="punctuation">,</span> dims<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">30</span><span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> FindClusters<span class="punctuation">(</span>sce<span class="punctuation">,</span> resolution<span class="operator">=</span><span class="number">0.5</span><span class="punctuation">,</span> verbose<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> RunTSNE<span class="punctuation">(</span>sce<span class="punctuation">,</span> dims<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">30</span><span class="punctuation">)</span>;</span><br><span class="line">sce <span class="operator">=</span> RunUMAP<span class="punctuation">(</span>sce<span class="punctuation">,</span> dims<span class="operator">=</span><span class="number">1</span><span class="operator">:</span><span class="number">30</span><span class="punctuation">,</span> verbose<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br></pre></td></tr></table></figure>

<h2 id="可视化"><a href="#可视化" class="headerlink" title="可视化"></a>可视化</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## show No. of cells per cluster</span></span><br><span class="line">ggplot<span class="punctuation">(</span>count.before.DF<span class="punctuation">,</span> aes<span class="punctuation">(</span>Var1<span class="punctuation">,</span> Freq<span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">+</span>geom_col<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">labs<span class="punctuation">(</span>title<span class="operator">=</span>paste0<span class="punctuation">(</span><span class="string">&quot;Before doublet removal, cells=&quot;</span><span class="punctuation">,</span> total.before.DF<span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">geom_text<span class="punctuation">(</span> aes<span class="punctuation">(</span>label<span class="operator">=</span>Freq<span class="punctuation">)</span><span class="punctuation">,</span> nudge_y<span class="operator">=</span><span class="number">50</span> <span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line">count.after.DF <span class="operator">=</span> as.data.frame<span class="punctuation">(</span>table<span class="punctuation">(</span>sce<span class="operator">@</span>meta.data<span class="operator">$</span>RNA_snn_res.0.5<span class="punctuation">)</span><span class="punctuation">)</span>;</span><br><span class="line">ggplot<span class="punctuation">(</span>count.after.DF<span class="punctuation">,</span> aes<span class="punctuation">(</span>Var1<span class="punctuation">,</span> Freq<span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">+</span>geom_col<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">labs<span class="punctuation">(</span>title<span class="operator">=</span>paste0<span class="punctuation">(</span><span class="string">&quot;After doublet removal, cells=&quot;</span><span class="punctuation">,</span> <span class="built_in">dim</span><span class="punctuation">(</span>sce<span class="punctuation">)</span><span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">geom_text<span class="punctuation">(</span> aes<span class="punctuation">(</span>label<span class="operator">=</span>Freq<span class="punctuation">)</span><span class="punctuation">,</span> nudge_y<span class="operator">=</span><span class="number">30</span> <span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line"><span class="comment">## show the UMI/feature/MT numbers and phase per cluster</span></span><br><span class="line">pa <span class="operator">=</span> FeaturePlot<span class="punctuation">(</span>sce<span class="punctuation">,</span> reduction<span class="operator">=</span><span class="string">&quot;umap&quot;</span><span class="punctuation">,</span> features<span class="operator">=</span><span class="string">&#x27;nCount_RNA&#x27;</span><span class="punctuation">)</span>;</span><br><span class="line">pb <span class="operator">=</span> FeaturePlot<span class="punctuation">(</span>sce<span class="punctuation">,</span> reduction<span class="operator">=</span><span class="string">&quot;umap&quot;</span><span class="punctuation">,</span> features<span class="operator">=</span><span class="string">&#x27;nFeature_RNA&#x27;</span><span class="punctuation">)</span>;</span><br><span class="line">pc <span class="operator">=</span> FeaturePlot<span class="punctuation">(</span>sce<span class="punctuation">,</span> reduction<span class="operator">=</span><span class="string">&quot;umap&quot;</span><span class="punctuation">,</span> features<span class="operator">=</span><span class="string">&#x27;percent.mito&#x27;</span><span class="punctuation">)</span>;</span><br><span class="line">pd <span class="operator">=</span> DimPlot<span class="punctuation">(</span>sce<span class="punctuation">,</span> reduction<span class="operator">=</span><span class="string">&quot;umap&quot;</span><span class="punctuation">,</span> group.by<span class="operator">=</span><span class="string">&quot;Phase&quot;</span><span class="punctuation">)</span>;</span><br><span class="line">grid.arrange<span class="punctuation">(</span>pa<span class="punctuation">,</span> pb<span class="punctuation">,</span> pc<span class="punctuation">,</span> pd<span class="punctuation">,</span> nrow<span class="operator">=</span><span class="number">2</span><span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line"><span class="comment">## Plot the cluster identifications</span></span><br><span class="line">TSNEPlot<span class="punctuation">(</span>sce<span class="punctuation">,</span> pt.size<span class="operator">=</span><span class="number">.5</span><span class="punctuation">,</span> label<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span> group.by<span class="operator">=</span><span class="string">&#x27;RNA_snn_res.0.5&#x27;</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">ggtitle<span class="punctuation">(</span><span class="string">&#x27;res=0.5, after doublet removal&#x27;</span><span class="punctuation">)</span>;</span><br><span class="line">UMAPPlot<span class="punctuation">(</span>sce<span class="punctuation">,</span> pt.size<span class="operator">=</span><span class="number">.5</span><span class="punctuation">,</span> label<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span> group.by<span class="operator">=</span><span class="string">&#x27;RNA_snn_res.0.5&#x27;</span><span class="punctuation">)</span><span class="operator">+</span></span><br><span class="line">ggtitle<span class="punctuation">(</span><span class="string">&#x27;res=0.5, after doublet removal&#x27;</span><span class="punctuation">)</span>;</span><br><span class="line"></span><br><span class="line">sce <span class="operator">=</span> SetIdent<span class="punctuation">(</span>sce<span class="punctuation">,</span>value<span class="operator">=</span><span class="string">&#x27;RNA_snn_res.0.5&#x27;</span><span class="punctuation">)</span>;</span><br><span class="line">sce<span class="operator">$</span>seurat_clusters<span class="operator">=</span>sce<span class="operator">$</span>RNA_snn_res.0.5;</span><br><span class="line"></span><br><span class="line"><span class="comment">## find cluster markers</span></span><br><span class="line">cm <span class="operator">=</span> FindAllMarkers<span class="punctuation">(</span>sce<span class="punctuation">,</span> logfc.threshold<span class="operator">=</span><span class="number">0.2</span><span class="punctuation">,</span> min.pct<span class="operator">=</span><span class="number">0.25</span><span class="punctuation">,</span> only.pos<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span> verbose<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">)</span>;</span><br><span class="line">cm <span class="operator">=</span> cm<span class="punctuation">[</span>with<span class="punctuation">(</span>cm<span class="punctuation">,</span>order<span class="punctuation">(</span>cluster<span class="punctuation">,</span> p_val_adj<span class="punctuation">,</span> <span class="operator">-</span>avg_logFC<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">,</span><span class="punctuation">]</span>;</span><br><span class="line"></span><br><span class="line"><span class="comment">## plot known markers</span></span><br><span class="line">FeaturePlot<span class="punctuation">(</span>sce<span class="punctuation">,</span> reduction<span class="operator">=</span><span class="string">&quot;umap&quot;</span><span class="punctuation">,</span></span><br><span class="line">			features<span class="operator">=</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;EPCAM&quot;</span><span class="punctuation">,</span> <span class="string">&quot;PTPRC&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD3D&quot;</span><span class="punctuation">,</span> <span class="string">&quot;CD68&quot;</span><span class="punctuation">)</span> <span class="punctuation">)</span>;</span><br></pre></td></tr></table></figure>

<h2 id="保存"><a href="#保存" class="headerlink" title="保存"></a>保存</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## save project</span></span><br><span class="line">save<span class="punctuation">(</span>sce<span class="punctuation">,</span> file<span class="operator">=</span>paste0<span class="punctuation">(</span>sid<span class="punctuation">,</span> <span class="string">&quot;.rds&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span>;</span><br><span class="line">write.csv<span class="punctuation">(</span>cm<span class="punctuation">,</span> file<span class="operator">=</span>paste0<span class="punctuation">(</span>sid<span class="punctuation">,</span> <span class="string">&quot;.cluster.markers.csv&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span>;</span><br></pre></td></tr></table></figure>



</div><div class="article-licensing box"><div class="licensing-title"><p>scRNA分析流程（一）</p><p><a href="http://example.com/2023/11/11/scRNA_1/">http://example.com/2023/11/11/scRNA_1/</a></p></div><div class="licensing-meta level is-mobile"><div class="level-left"><div class="level-item is-narrow"><div><h6>Author</h6><p>Se</p></div></div><div class="level-item is-narrow"><div><h6>Posted on</h6><p>2023-11-11</p></div></div><div class="level-item is-narrow"><div><h6>Updated on</h6><p>2023-11-11</p></div></div><div class="level-item is-narrow"><div><h6>Licensed under</h6><p><a class="icons" rel="noopener" target="_blank" title="Creative Commons" href="https://creativecommons.org/"><i class="icon fab fa-creative-commons"></i></a><a class="icons" rel="noopener" target="_blank" title="Attribution" href="https://creativecommons.org/licenses/by/4.0/"><i class="icon fab fa-creative-commons-by"></i></a><a class="icons" rel="noopener" target="_blank" title="Noncommercial" href="https://creativecommons.org/licenses/by-nc/4.0/"><i class="icon fab fa-creative-commons-nc"></i></a></p></div></div></div></div></div><div class="article-tags is-size-7 mb-4"><span class="mr-2">#</span><a class="link-muted mr-2" rel="tag" href="/tags/Pipeline/">Pipeline</a><a class="link-muted mr-2" rel="tag" href="/tags/scRNA/">scRNA</a></div><!--!--></article></div><!--!--><nav class="post-navigation mt-4 level is-mobile"><div class="level-end"><a class="article-nav-next level level-item link-muted" href="/2023/10/26/RNAseq/"><span class="level-item">RNAseq 分析流程</span><i class="level-item fas fa-chevron-right"></i></a></div></nav><div class="card"><div class="card-content"><h3 class="title is-5">Comments</h3><div id="comment-container"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/gitalk@1.7.2/dist/gitalk.css"><script src="https://cdn.jsdelivr.net/npm/gitalk@1.7.2/dist/gitalk.min.js"></script><script>var gitalk = new Gitalk({
            id: "ae35beb2bf604e1fe00b69fb82f2d643",
            repo: "gitalk_comments",
            owner: "iAMSe",
            clientID: "c7f7a9ad7a6bb47dbaeb",
            clientSecret: "eecbf00167ee422545a5dc5bbafc7142bf719a98",
            admin: ["iAMSe"],
            createIssueManually: false,
            distractionFreeMode: false,
            perPage: 20,
            pagerDirection: "last",
            
            
            enableHotKey: true,
            language: "zh-CN",
        })
        gitalk.render('comment-container')</script></div></div></div><div class="column column-left is-4-tablet is-4-desktop is-4-widescreen  order-1 is-sticky"><div class="card widget" data-type="profile"><div class="card-content"><nav class="level"><div class="level-item has-text-centered flex-shrink-1"><div><figure class="image is-128x128 mx-auto mb-2"><img class="avatar is-rounded" src="/imgs/dinosaur.jpg" alt="小傻的帽子🎓"></figure><p class="title is-size-4 is-block" style="line-height:inherit;">小傻的帽子🎓</p><p class="is-size-6 is-block">小傻帽和小傻子</p></div></div></nav><nav class="level is-mobile"><div class="level-item has-text-centered is-marginless"><div><p class="heading">Posts</p><a href="/archives"><p class="title">4</p></a></div></div><div class="level-item has-text-centered is-marginless"><div><p class="heading">Category</p><a href="/categories"><p class="title">1</p></a></div></div><div class="level-item has-text-centered is-marginless"><div><p class="heading">Tags</p><a href="/tags"><p class="title">6</p></a></div></div></nav></div></div><div class="card widget" id="toc" data-type="toc"><div class="card-content"><div class="menu"><h3 class="menu-label">Catalogue</h3><ul class="menu-list"><li><a class="level is-mobile" href="#数据处理"><span class="level-left"><span class="level-item">1</span><span class="level-item">数据处理</span></span></a><ul class="menu-list"><li><a class="level is-mobile" href="#Smart-seq数据"><span class="level-left"><span class="level-item">1.1</span><span class="level-item">Smart-seq数据</span></span></a></li><li><a class="level is-mobile" href="#10×数据处理"><span class="level-left"><span class="level-item">1.2</span><span class="level-item">10×数据处理</span></span></a></li></ul></li><li><a class="level is-mobile" href="#Seurat-处理"><span class="level-left"><span class="level-item">2</span><span class="level-item">Seurat 处理</span></span></a><ul class="menu-list"><li><a class="level is-mobile" href="#加载数据"><span class="level-left"><span class="level-item">2.1</span><span class="level-item">加载数据</span></span></a><ul class="menu-list"><li><a class="level is-mobile" href="#Smart-seq"><span class="level-left"><span class="level-item">2.1.1</span><span class="level-item">Smart seq</span></span></a></li><li><a class="level is-mobile" href="#10×"><span class="level-left"><span class="level-item">2.1.2</span><span class="level-item">10×</span></span></a></li></ul></li><li><a class="level is-mobile" href="#质控"><span class="level-left"><span class="level-item">2.2</span><span class="level-item">质控</span></span></a></li><li><a class="level is-mobile" href="#标准化-PCA"><span class="level-left"><span class="level-item">2.3</span><span class="level-item">标准化 PCA</span></span></a></li><li><a class="level is-mobile" href="#可视化"><span class="level-left"><span class="level-item">2.4</span><span class="level-item">可视化</span></span></a></li><li><a class="level is-mobile" href="#保存"><span class="level-left"><span class="level-item">2.5</span><span class="level-item">保存</span></span></a></li></ul></li></ul></div></div><style>#toc .menu-list > li > a.is-active + .menu-list { display: block; }#toc .menu-list > li > a + .menu-list { display: none; }</style><script src="/js/toc.js" defer></script></div><div class="card widget" data-type="recent-posts"><div class="card-content"><h3 class="menu-label">Recents</h3><article class="media"><div class="media-content"><p class="date"><time dateTime="2023-11-11T13:15:05.386Z">2023-11-11</time></p><p class="title"><a href="/2023/11/11/scRNA_1/">scRNA分析流程（一）</a></p><p class="categories"><a href="/categories/pipeline/">pipeline</a></p></div></article><article class="media"><div class="media-content"><p class="date"><time dateTime="2023-10-26T11:11:52.699Z">2023-10-26</time></p><p class="title"><a href="/2023/10/26/RNAseq/">RNAseq 分析流程</a></p><p class="categories"><a href="/categories/pipeline/">pipeline</a></p></div></article><article class="media"><div class="media-content"><p class="date"><time dateTime="2023-10-13T10:27:26.233Z">2023-10-13</time></p><p class="title"><a href="/2023/10/13/Msuite2/">Msuite2:一体式DNA甲基化分析及可视化软件</a></p><p class="categories"><a href="/categories/pipeline/">pipeline</a></p></div></article><article class="media"><div class="media-content"><p class="date"><time dateTime="2023-10-13T09:37:06.030Z">2023-10-13</time></p><p class="title"><a href="/2023/10/13/hello-world/">Hello World</a></p></div></article></div></div><div class="card widget" data-type="tags"><div class="card-content"><div class="menu"><h3 class="menu-label">Tags</h3><div class="field is-grouped is-grouped-multiline"><div class="control"><a class="tags has-addons" href="/tags/Epigenome/"><span class="tag">Epigenome</span><span class="tag">1</span></a></div><div class="control"><a class="tags has-addons" href="/tags/Methylation/"><span class="tag">Methylation</span><span class="tag">1</span></a></div><div class="control"><a class="tags has-addons" href="/tags/Pipeline/"><span class="tag">Pipeline</span><span class="tag">3</span></a></div><div class="control"><a class="tags has-addons" href="/tags/Transcriptome/"><span class="tag">Transcriptome</span><span class="tag">1</span></a></div><div class="control"><a class="tags has-addons" href="/tags/scRNA/"><span class="tag">scRNA</span><span class="tag">1</span></a></div><div class="control"><a class="tags has-addons" href="/tags/software/"><span class="tag">software</span><span class="tag">1</span></a></div></div></div></div></div></div><!--!--></div></div></section><footer class="footer"><div class="container"><div class="level"><div class="level-start"><a class="footer-logo is-block mb-2" href="/"><img src="/imgs/dinosaur.jpg" alt="Hat&#039;s blog" height="28"></a><p class="is-size-7"><span>&copy; 2023 Se</span>  Powered by <a href="https://hexo.io/" target="_blank" rel="noopener">Hexo</a> &amp; <a href="https://github.com/ppoffice/hexo-theme-icarus" target="_blank" rel="noopener">Icarus</a></p></div><div class="level-end"></div></div></div></footer><script src="https://cdn.jsdelivr.net/npm/jquery@3.3.1/dist/jquery.min.js"></script><script src="https://cdn.jsdelivr.net/npm/moment@2.22.2/min/moment-with-locales.min.js"></script><script src="https://cdn.jsdelivr.net/npm/clipboard@2.0.4/dist/clipboard.min.js" defer></script><script>moment.locale("en");</script><script>var IcarusThemeSettings = {
            article: {
                highlight: {
                    clipboard: true,
                    fold: 'unfolded'
                }
            }
        };</script><script src="/js/column.js"></script><script src="/js/animation.js"></script><a id="back-to-top" title="Back to top" href="javascript:;"><i class="fas fa-chevron-up"></i></a><script src="/js/back_to_top.js" defer></script><!--!--><!--!--><!--!--><script src="https://cdn.jsdelivr.net/npm/cookieconsent@3.1.1/build/cookieconsent.min.js" defer></script><script>window.addEventListener("load", () => {
      window.cookieconsent.initialise({
        type: "info",
        theme: "edgeless",
        static: false,
        position: "bottom-left",
        content: {
          message: "This website uses cookies to improve your experience.",
          dismiss: "Got it!",
          allow: "Allow cookies",
          deny: "Decline",
          link: "Learn more",
          policy: "Cookie Policy",
          href: "https://www.cookiesandyou.com/",
        },
        palette: {
          popup: {
            background: "#edeff5",
            text: "#838391"
          },
          button: {
            background: "#4b81e8"
          },
        },
      });
    });</script><script src="https://cdn.jsdelivr.net/npm/lightgallery@1.10.0/dist/js/lightgallery.min.js" defer></script><script src="https://cdn.jsdelivr.net/npm/justifiedGallery@3.8.1/dist/js/jquery.justifiedGallery.min.js" defer></script><script>window.addEventListener("load", () => {
            if (typeof $.fn.lightGallery === 'function') {
                $('.article').lightGallery({ selector: '.gallery-item' });
            }
            if (typeof $.fn.justifiedGallery === 'function') {
                if ($('.justified-gallery > p > .gallery-item').length) {
                    $('.justified-gallery > p > .gallery-item').unwrap();
                }
                $('.justified-gallery').justifiedGallery();
            }
        });</script><!--!--><!--!--><!--!--><!--!--><!--!--><script src="/js/main.js" defer></script><div class="searchbox"><div class="searchbox-container"><div class="searchbox-header"><div class="searchbox-input-container"><input class="searchbox-input" type="text" placeholder="Type something..."></div><a class="searchbox-close" href="javascript:;">×</a></div><div class="searchbox-body"></div></div></div><script src="/js/insight.js" defer></script><script>document.addEventListener('DOMContentLoaded', function () {
            loadInsight({"contentUrl":"/content.json"}, {"hint":"Type something...","untitled":"(Untitled)","posts":"Posts","pages":"Pages","categories":"Categories","tags":"Tags"});
        });</script></body></html>